### ISSUE TITLE
some questions

### ISSUE THREAD

**Issue by zh583007354** [Ziems/ProxylessNAS#2]
Hello, 
I am working hard to reimplement this ProxylessNAS, recently. 
Your project attracts my attention, but I have some question.
Firstly, you have a new method to reimplement it. The different child graphs(or networks) are individual, and the parent graph don't exist factly. I never thought to be able to do so, but still try to create a parent graph and sample the child graph by model.apply()......

1.when updating the weights of Network, you use the AdamOptimizer on it. Although there is no mention about the optimizer of weights (I think), accroding to the expreience of DARTS, should we use the SGD for the weight parameters?

2. According to the sec.3.2.1, "when training weight parameters, we first freeze the architecture parameters and stochastically sample binary gates accroding to Eq.2 for each batch", I think that maybe you should use choice(len(), p=probs) in train_layer(). 

3.Does the current code need to use muti-gpu manually?

4.when updating the architecture params, in the line 147 of trainer.py,
`dLdg_b = torch.sum(block_a.block_grad * block_b.block_forward`).item()
I think the variable dLdg_b should be computed by the block_a.block_grad * block_b.block_forward, rather than block_b.block_grad. 
Because dL/dg_b = dL/dy * dy/dg_b = dL/dy * o_b(x) accroding to the chin rule,
y = g_a * o_a(x) + g_b * o_b(x) = o_a(x), where g_a = 1 and g_b = 0.

Maybe you can teach me why you  use the block_b.block_grad in your code if my reasoning is wrong.

5.Last question is that  your code update the architecture params of randomly choiced one layer, but all layers during one iteration step. Is this your personly understanding, or can it only be implemented this way? I will ask this because the DARTS algorithm can update all layers alpha params one step.

Hoping your reply.
Good job!

**Comment by Ziems** [comment]
You are correct. I actually had quite a bit of trouble understanding the parent graphs, and I emailed the writers of ProxylessNAS to get a better explanation but none of them responded.

1. Ah okay! Lets try it!

2. Yes I did add this to the project, but I did not commit my changes and my computer crashed so I need to re-implement them. It took a very small amount of time though.

3. The current code was written to run on the cluster server that my university has. I could not find an easier way to do computations in parallel, but I am a complete beginner and I would appreciate any help at all!

4. So the block_a and block_b are two separate computation blocks that are interchangeable. What I do is I compute the loss of the network with block_a, then swap it out for block_b and do the same. Then, I calculate the sum of the loss of each block and that becomes the blocks loss. At no point do the separate block do any computation together, so their gradients should not intermingle like that. At least that's my understanding.

**Comment by zh583007354** [comment]
Hi, boy. I have reimplemented a version with construting a parent graph by ModuleList() and sampling by model.apply().

But, because the information security of my company, I can't upload my code now. Actually, my code is not completed about the complex experimential settings and It can only run with my own understanding. Maybe I will upload them to my repository from my own PC in a few days.

**Comment by Ziems** [comment]
Sure! Feel free to make a pull request!

**Comment by xieydd** [comment]
@zh583007354  Still WIP?

**Comment by xieydd** [comment]
Official training code is released， but sadly not include search phase. [ProxylessNAS](https://github.com/mit-han-lab/ProxylessNAS)
